엔티티 연관관계 설정 변경

kimcno3 opened the pull request:
* 외래키를 가진 객체의 연관관계 객체 설정에 insertable, updatable 값을 false로 변경

* 외래키를 직접 저장할 외래키 필드 추가(updatable은 false로 지정)

* 변경된 엔티티 객체를 save하는 로직에서 연관객체 조회 로직 제거(불필요한 select문 발생 지점 삭제)

kimcno3 committed:
엔티티 연관관계 설정 변경

* 외래키를 가진 객체의 연관관계 객체 설정에 insertable, updatable 값을 false로 변경

* 외래키를 직접 저장할 외래키 필드 추가(updatable은 false로 지정)

* 변경된 엔티티 객체를 save하는 로직에서 연관객체 조회 로직 제거(불필요한 select문 발생 지점 삭제)

kimcno3 commented:
외래키를 저장하는 필드를 추가하면서 엔티티 객체 저장시, 연관관계 객체를 조회하는 로직을 제거하도록 구성했습니다!

f-lab-bright commented on src/main/java/com/flab/football/service/match/MatchServiceImpl.java:
```diff
@@ -66,47 +35,30 @@ public void createMatch(CreateMatchCommand command) {
         .shoes(command.getShoes())
         .level(command.getLevel())
         .gender(command.getGender())
+        .stadiumId(command.getStadiumId())
```
이 id가 존재하지 않거나 잘못된 값이 들어오면 어떻게 될까요?

f-lab-bright commented on src/main/java/com/flab/football/service/match/MatchServiceImpl.java:
```diff
@@ -66,47 +35,30 @@ public void createMatch(CreateMatchCommand command) {
+        .build();
+
+    Match.Manager manager = Manager.builder()
+        .userId(command.getUserId())
```
이 userId가 존재하지 않거나 잘못된 값이라면..?

f-lab-bright commented on src/main/java/com/flab/football/service/match/MatchServiceImpl.java:
```diff
@@ -66,47 +35,30 @@ public void createMatch(CreateMatchCommand command) {
     Match.Member member = Match.Member.builder()
-        .user(user.get())
-        .match(match.get())
+        .userId(userId)
+        .matchId(matchId)
```
이 값들도 존재하지 않거나 잘못된 값이라면..?

f-lab-bright commented on src/main/java/com/flab/football/domain/Match.java:
```diff
@@ -93,8 +93,16 @@ public enum LimitGender {
+      referencedColumnName = "id",
+      insertable = false,
+      updatable = false
+  )
```
Match entity에서 다른 연관 객체들은 변경할 일이 전혀 없고, 조회를 위해서만 연관성을 가지고 있는건가요??

kimcno3 committed:
피드백 반영영

* 외래키 값을 직접 저장하는 필드에 insertable, updatable 값을 false로 변경
  - 외래키 필드는 순수하게 조회의 목적으로만 사용하기 위한 변경

* 연관관계 객체 필드에 insertable, updatable 값은 true로 변경
  - 외래키를 직접 입력하는 것이 아닌 연관관계 객체를 통해 다른 테이블의 정보도 저장 및 수정되도록 구조 변경

* 엔티티 객체를 save하는 로직에서 연관 객체 조회를 위한 메소드 다시 추가
  - 불필요한 작업이 아닌 외래키에 대한 검증과정을 거치는 것.
  - 이를 제거하고 외래키를 그대로 객체에 저장한다면 검증에 대한 절차를 생략하는 것이다.

* 연관관계 객체  조회 메소드 호출시 orElseThrow() 사용

kimcno3 commented:
성능 최적화에 너무 포커스를 맞춰 외래키 검증에 대한 절차를 모두 제거한 것 같습니다.

그리고 연관관계 생성 및 저장하는 과정에서 연관관계 객체를 조회하는 로직은 다시 추가하고 객체간의 연결을 통해 엔티티 객체의 정보를 구성하도록 설정을 바꿨습니다.

그리고 외래키 값을 저장하는 필드는 오직 외래키 정보를 조회하는 목적으로만 사용하기 위해  insertable, updatable 을 false로 설정했습니다!

f-lab-bright reviewed (approved):
LGTM with Comments
command 요청에 대해서 실제 값이 있는지 확인하기 위한 조회 요청은 너무 많지 않다면 부담없이 하셔도 됩니다.
다만, 이전에 insert 이후에 update가 일어나는 등의 실제로 불필요한 쿼리들만 실행되지 않도록 해두시면 충분합니다 :)

kimcno3 commented:
네!!
